@here/harp-mapview/lib/ClipPlanesEvaluator.ts

```diff
@@ -0,0 +1,524 @@
+     * [[nearMargin]] + [[farMargin]] can not be bigger then maximum offset between a planes
+     * assumed - [[planesMaxDistance]].
+     * @note [[planesMaxDistance]] does not limit far plane when spherical projection is in use,
+     * the algorithm used estimates distance to point on tangent where line from camera touches
+     * the sphere horizont.
+     * @param nearMargin defines near plane offset from the ground in the surface normal direction.
+     * @param farMargin defines far plane offset from the ground surface, looking downwards.
+     * @param nearMin minimum allowable near plance distance from camera.
+     * @param planesMaxDistance maximum offset between near/far planes, allows to limit viewing
+     * distance at overall.
+     */
+    constructor(
+        public nearMargin: number = 200,
+        public farMargin: number = 50,
+        readonly nearMin: number = 0.1,
+        readonly planesMaxDistance = 100000.0
+    ) {
+        assert(nearMargin >= 0);
+        assert(farMargin >= 0);
+        assert(nearMin > 0);
+        assert(planesMaxDistance >= 0);
+        assert(nearMargin + farMargin <= planesMaxDistance);
+    }
+
+    evaluateClipPlanes(
+        camera: THREE.Camera,
+        projection: Projection
+    ): { near: number; far: number } {
+        const clipPlanes =
+            projection.type !== ProjectionType.Spherical
+                ? this.evaluateDistancePlanarProj(camera, projection)
+                : this.evaluateDistanceSphericalProj(camera, projection);
+
+        return clipPlanes;
+    }
+
+    /**
+     * Calculate camera altitute (closest distance) to ground level in world units.
+     * @param camera
+     * @param projection
+     */
+    protected getCameraAltitude(camera: THREE.Camera, projection: Projection) {
+        return projection.groundDistance(camera.position);
+    }
+
+    private evaluateDistancePlanarProj(
+        camera: THREE.Camera,
+        projection: Projection
+    ): { near: number; far: number } {
+        assert(projection.type !== ProjectionType.Spherical);
+
+        let nearPlane: number = this.nearMin;
+        let farPlane: number = this.nearMin + this.planesMaxDistance;
+
+        // Calculate distance to closest point on the ground.
+        const groundDistance = this.getCameraAltitude(camera, projection);
+        // Acquire camera forward (look at) vector (store in tmpVectors[0]).
+        const lookAt = camera.getWorldDirection(this.m_tmpVectors[0]);
+        const tiltCos = Math.cos(
+            lookAt.angleTo(AltitudeBasedClipPlanesEvaluator.GROUND_NORMAL_INV_PLANAR_PROJ)
+        );
+        // At least try to keep margins along the eye vector (center of the view) in tact with
+        // pitch angle changes - this does not solve all tilt change problems, rather use
+        // more sophisticated evaluator.
+        nearPlane = groundDistance - this.nearMargin / tiltCos;
```

Comment:
Do you need to check for division by zero?

Reply:
That's good point! Will fix of course.